keps/sig-api-machinery/20190329-less-object-serializations.md

```diff
@@ -0,0 +1,258 @@
+
+**Note:** Any PRs to move a KEP to `implementable` or significant changes once it is marked `implementable` should be approved by each of the KEP approvers. If any of those approvers is no longer appropriate than changes to that list should be approved by the remaining approvers and/or the owning SIG (or SIG-arch for cross cutting KEPs).
+
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://github.com/kubernetes/enhancements/issues
+[kubernetes/kubernetes]: https://github.com/kubernetes/kubernetes
+[kubernetes/website]: https://github.com/kubernetes/website
+
+## Summary
+
+Scalability and performance of kube-apiserver is crucial for scalability
+of the whole Kubernetes cluster. Given that kube-apiserver is cpu-intensive
+process, scaling a single instance of it translates to optimizing amount
+of work is needed to process a request (cpu cycles and amount of allocated
+memory, as memory management is significant part of work done be
+kube-apiserver).
+
+This proposal is aiming to significantly reduce amount of work spent on
+serializing objects as well as amount of allocated memory to process that.
+
+## Motivation
+
+Running different types of scalability tests and analyzing large production
+clusters proves that large number of watches watching the same set of objects
+may cause significant load on kube-apiserver. An extreme example of it is
+[#75294][], where creation of a single large Endpoints object (almost 1MB of
+size, due to 5k pods backing it) in 5k-node cluster can completely overload
+kube-apiserver for 5 seconds.
+
+The main reason for that is that for every watcher (Endpoints are being watched
+by kube-proxy running on every one) kube-apiserver independently serializes
+(which also requires deep-copy) every single object being send via this watch.
+
+While this problem is extremely visible for watch, the situation looks the same
+for regular GET/LIST operations - reading the same object N times will result
+in serializing that N times independently.
+
+This proposal presents a solution for that problem.
+
+[#75294]: https://github.com/kubernetes/kubernetes/issues/75294
+
+### Goals
+
+- Reduce load on kube-apiserver, by not serializing the same object multiple
+times wherever it is possible.
+
+### Non-Goals
+
+- Change overall architecture of the system, by changing what data is being
+read/watched by different components.
+
+## Proposal
+
+This proposal does not introduce any user-visible changes - the proposed changes
+are purely implementational detail of kube-apiserver.
+
+We propose introducing an internal type implementing `runtime.Object` interface,
+that is encapsulating the original object and is additionally able to accumulate
+its serialized versions. It will look like this:
```

Comment:
Another idea for an alternative: modify / add to the serialization code such that
* proto objects keep the raw bytes; a metadata accessor is generated that either reads the raw proto bytes or deserializes only metadata.
* Setting self link and RV shuffles bytes around rather than decoding and re-encoding.

This would be more work but I'm willing to bet that the performance gains would be 2x as big. Additionally, this should save memory overall rather than consume additional memory, since the deserialized form never needs to be produced.

I also feel like this would be less invasive--I can probably live with the design currently written in the doc, but I'm not excited about the distributed implicit cache, it adds a lot of non-local state.

Reply:
> Another idea for an alternative: modify / add to the serialization code such that
> 
> * proto objects keep the raw bytes; a metadata accessor is generated that either reads the raw proto bytes or deserializes only metadata.
> * Setting self link and RV shuffles bytes around rather than decoding and re-encoding.

This sounds similar to the smart objects discussion we've had in the past.  I'm interested in the idea of smart objects that only serialize or deserialize on demand, but the effort is significant.

Reply:
> proto objects keep the raw bytes; a metadata accessor is generated that either reads the raw proto bytes or deserializes only metadata.

Metadata is not enough - as an example for pods, we are filtering by NodeName.

> Setting self link and RV shuffles bytes around rather than decoding and re-encoding.

That's quite tricky and error prone for protobufs, I'm afraid.

One thing that we're keep forgetting are different object versions - from etcd we get object in version X, it doesn't mean that it is being served to watcher also in version X. I think making that assumption may cause us significant issues, when we e.g. will decide to migrate to new version of Endpoints object (which is under discussion).
So I'm quite reluctant to this option.

Reply:
Yeah, that's a good point. The proto munging would only be available in one case, although it's likely to be a common case.